MeshGeoToolsLib/BoundaryElementsAlongPolyline.cpp

```diff
@@ -48,27 +48,61 @@ BoundaryElementsAlongPolyline::BoundaryElementsAlongPolyline(
             auto* edge = e->getEdge(i);
             // check if all edge nodes are along the polyline (if yes, store a distance)
             std::vector<std::size_t> edge_node_distances_along_ply;
-            if (includesAllEdgeNodeIDs(node_ids_on_poly, *edge, edge_node_distances_along_ply)) {
-                auto* new_edge = modifyEdgeNodeOrdering(*edge, ply, edge_node_distances_along_ply, node_ids_on_poly);
+            if (includesAllEdgeNodeIDs(node_ids_on_poly, *edge,
+                                       edge_node_distances_along_ply))
+            {
+                auto* new_edge = modifyEdgeNodeOrdering(
+                    *edge, ply, edge_node_distances_along_ply,
+                    node_ids_on_poly);
                 if (edge != new_edge)
                     delete edge;
                 _boundary_elements.push_back(new_edge);
+                _bulk_ids.emplace_back(e->getID(), i);
             } else {
                 delete edge;
             }
         }
     }
 
-    // sort picked edges according to a distance of their first node along the polyline
-    std::sort(_boundary_elements.begin(), _boundary_elements.end(),
-            [&](MeshLib::Element*e1, MeshLib::Element*e2)
-            {
-                std::size_t dist1 = std::distance(node_ids_on_poly.begin(),
-                    std::find(node_ids_on_poly.begin(), node_ids_on_poly.end(), e1->getNodeIndex(0)));
-                std::size_t dist2 = std::distance(node_ids_on_poly.begin(),
-                    std::find(node_ids_on_poly.begin(), node_ids_on_poly.end(), e2->getNodeIndex(0)));
-                return (dist1 < dist2);
-            });
+    // zip _boundary_elements and _bulk_element_ids
```

Comment:
Very similar algorithm zip-sort-unzip is used in BaseLib::quicksort. Maybe the BaseLib version can be generalized and used for this purpose? The only difference there is that `begin` is a pointer and offset, and same for the `end`.